docs/config.rst
=======================
Configuration Reference
=======================

Stages, such as *dev*, *staging*, and *production* are configured in the *zappa_settings.json* file.  Here is an example of defining many of the available settings:

::

    {
        "dev": {
            "api_key_required": false, // enable securing API Gateway endpoints with x-api-key header (default False)
            "assume_policy": "my_assume_policy.json", //
            "attach_policy": "my_attach_policy.json", // optional, IAM attach policy JSON file
            "authorization_type": "NONE", // optional, use "AWS_IAM" to require request signing
            "aws_region": "us-east-1",
            "cache_cluster_enabled": false,
            "cache_cluster_size": .5,
            "callbacks": { // Call custom functions during the local Zappa deployment/update process
                "settings": "my_app.settings_callback", // After loading the settings
                "zip": "my_app.zip_callback", // After creating the package
                "post": "my_app.post_callback", // After command has executed
            },
            "dead_letter_arn": "arn:aws:<sns/sqs>:::my-topic/queue", // Optional Dead Letter configuration for when Lambda async invoke fails thrice
            "debug": true
            "delete_zip": true
            "domain": "yourapp.yourdomain.com",
           "events": [
                {   // Recurring events
                    "function": "your_module.your_recurring_function", // The function to execute
                    "expression": "rate(1 minute)" // When to execute it (in cron or rate format)
                },
                {   // AWS Reactive events
                    "function": "your_module.your_reactive_function", // The function to execute
                    "event_source": {
                        "arn":  "arn:aws:s3:::my-bucket", // The ARN of this event source
                        "events": [
                            "s3:ObjectCreated:*" // The specific event to execute in response to.
                        ]
                    }
                }
            ],
            "exclude": ["*.gz", "*.pem"],
            "http_methods": ["GET", "POST"],
            "integration_response_codes": [200, 301, 404, 500],
            "keep_warm": true,
            "lambda_handler": "your_custom_handler",
            "log_level": "DEBUG",
            "memory_size": 512,
            "method_response_codes": [200, 301, 404, 500],
            "parameter_depth": 10,
            "prebuild_script": "your_module.your_function",
            "profile_name": "your-profile-name",
            "project_name": "MyProject",
            "remote_env": "s3://my-project-config-files/filename.json",
                // optional file in s3 bucket containing a flat json object which will be used to set custom environment variables.
            "role_name": "MyLambdaRole",
            "s3_bucket": "dev-bucket",
            "settings_file": "~/Projects/MyApp/settings/dev_settings.py",
            "timeout_seconds": 30,
            "touch": false,
            "use_precompiled_packages": false,
            "use_apigateway": true,
            "vpc_config": {
                            "SubnetIds": [ "subnet-12345678" ],
                            "SecurityGroupIds": [ "sg-12345678" ]
                          }
        }    
    }


All values are standard JSON data types (Numbers, Strings, Booleans, Arrays, and Objects).

api_key_required
================

(Optional) This bool determines whether or not to enable securing API Gateway endpoints with x-api-key header (default False)

assume_policy
=============

(Optional) IAM assume policy JSON file

attach_policy
=============
(Optional) IAM attach policy JSON file

authorization_type
==================
(Optional) Default `NONE`. Use `AWS_IAM` to require signed requests

callbacks
=========

(Optional) Call custom functions during the local Zappa deployment/update process

aws_region
==========

This string setting specifies which AWS Region to use.

The default value is US East, which is *"us-east-1"*.

cache_cluster_enabled
=====================

This boolean setting indicates whether to use the APIGW cash cluster.

The default value is *false*.

cache_cluster_size
==================

This number setting specifies the APIGW Cache Cluster size.

The default value is *0.5*.

debug
=====

This boolean setting governs whether Zappa configuration errors tracebacks are to appear in HTTP 500 error pages.


delete_local_zip
==========

This boolean setting specifies whether to delete the local zip archive after code updates.

delete_s3_zip
==========

This boolean setting specifies whether to delete the s3 zip archive after code updates.

domain
======

This string setting is required if a domain will be used.

It should be in a format like *"yourapp.yourdomain.com"*.

events
======

This is an array with settings which describe the functions and schedules to execute them.

Each event should contain objects with values for *function* and *expression*?

function
--------

This string setting identifies the function being referenced in an event.

It should have a format like *"your_module.your_function"*.


expression
----------

This string setting provides an AWS Lambda schedule expression using Rate or Cron formats.  See the `AWS documentation <http://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html>`_ for a description of currently accepable formats for this setting.  This is the setting that defines when the function should be executed.

exclude
=======

This is an array of regex string patterns to exclude from the archive.

override_default_exclude
================

Some modules are excluded by default from the archive because they are already provided by AWS in the Lambda environment. This settings lets you provide a list of UNIX filename-style globs to filter that list. For example ``["boto*"]`` would prevent ``boto3`` from being excluded, and thus allow overriding the version that is provided by AWS (e.g. when it lags behind a version that's needed for a bug fix).

The default excluded modules for Python 2.x are:

- ``boto3``
- ``dateutil``
- ``botocore``
- ``s3transfer``
- ``six.py``
- ``jmespath``
- ``concurrent``

For Python 3+ they are:

- ``boto3``
- ``dateutil``
- ``botocore``
- ``s3transfer``
- ``concurrent``


http_methods
============

This array setting is a list of HTTP methods to route.  

Examples of HTTP methods are GET and POST, as in this example: *["GET", "POST"]*.


integration_response_codes
==========================

This is an array of integers which are integration response status codes to route.

This should in a formal like *[200, 301, 404, 500]*.

keep_warm
=========

This boolean setting is used to specify whether to create CloudWatch events to keep the server warm.

lambda_handler
==============

The string setting is the name of the Lambda handler.

The default is *"handler.lambda_handler"*.

log_level
=========

This string setting is used to set the Zappa log level.

The value of this setting can be either *"CRITICAL"*, *"ERROR"*, *"WARNING"*, *"INFO"* or *"DEBUG"*.  The default is *"DEBUG"*.

memory_size
===========

This number setting specifies the Lambda function memory in MB.

method_response_codes
=====================

This array setting is a list of method response status codes to route.

This should be in a format like *[200, 301, 404, 500]*.

parameter_depth
===============

This integer setting specifies the size of the URL depth to route.

This defaults to *8*.

prebuild_script
===============

This string setting identifies a function to execute before uploading code.

This should be in a format like *"your_module.your_function"*.

profile_name
============

This string setting identifies the profile name of the AWS credentials to use.

The default is *"default"*.

project_name
============

This string setting is the name of the project as it appears on AWS. 

It defaults to *a slugified `pwd`*.  <TODO: What does THAT mean??!>

role_name
=========

This string setting is name of the Lambda execution role.

s3_bucket
=========

This string setting is the name of the Zappa zip bucket.

settings_file
=============

This string setting is the full path for the server side settings file.

timeout_seconds
===============

This number setting specifies the maximum lifespan for the Lambda function in seconds.

The default is *30*.

touch
=====

This boolean setting determines whether to GET the production URL upon initial deployment.

Default is *true*.

use_precompiled_packages
========================

This boolean setting is used to indicate whether, if possible, to use the C-extension packages which have been pre-compiled for AWS Lambda.

use_apigateway
==============

This boolean setting indicates whether the API Gateway resource should be created.

The default is *true*.

vpc_config
==========

This setting provides some optional VPC configuration for Lambda function.  This value for this setting is an object with sub-settings.

SubnetsIds
----------

This is an array setting that is used to select subnets, which is a list of strings.

Note that not all availability zones support Lambda.


SecurityGroupIds
----------------

This is an array setting that is used to select security groups, which is a list of strings.


CHANGELOG.md
# Zappa Changelog

## 0.45.1
* Make `package_info.json` consistent across Python versions
* Fixes to `task_sns`
* Add `role_arn` settings variable

## 0.45.0
* Add support for SQLite on Python 3.6 (LP bump)
* Add unique package identifier file
* Fixes #1171 don't colorize partial words
* Various small changes
* More robust OpenSSL parsing for LE
* Update applies tags
* Allows deploying and updating from existing S3 package

## 0.44.3
* Slim handler packaging hotfix - thanks @mcrowson!

## 0.44.2
* Non-Dynamo asynx hotfix - thanks @jwkvam!

## 0.44.1
* Packaging hotfix - thank Yunseop!

## 0.44.0
* Async responses (thanks Sean!)
* Remove setLevel call from common log
* Fix #1023
* Merge #1058, 1059, 1073
* Allow overriding of endpoint_url arg
* Add support for API Gateway stage variables
* Add AWS X-Ray support (thanks @mathom!)
* Events take kwargs (Thanks @mcrowson!)
* Detect P2/3 during `init`
* Include stage names in `slim_handler` zips
* Allow `-s` with `manage`
* Use same S3 bucket for global endpoints
* Fix bug with mixed-case packages
* Cache wheels, fix 0-byte wheels

## 0.43.2
* Add index prefix to each subsequent schedule expression for the same event (#1051)
* fix/Only update CORS on resources which have an OPTIONS method (#1036)
* Support for binary support and cors simultaneously (#1011)
* Set Flask script name based on domain which the request was made to (#1015)
* Fix SNS Async (#1055)
* GZip for slim handler
* Force color option
* Various dep bumps

## 0.43.1
* Fixes #1001, don't override AWS env vars if k:v not set. Thanks Nik and Sean!

## 0.43.0
* Checks for the key 'Environment' when fetching remote lambda env vars (#988)
* except BotoCoreError before general exception in zappa.cli.update
* make cookie hack case-insensitive
* Fix #998 - Make environment variable keys strings instead of byte arrays in python 3.6
* Add --disable_progress command line parameter
* #946 - Allow setting cors to false.
* #870 Lambda from outside
* Implement context header mappings - Feature Request Issue #939
* Separating out native AWS environment variables ##962
* Rule name shortening
* Splintering aws_environment_variables from environment_variables (to avoid overwriting AWS native env vars).

## 0.42.2
* Add exclude for __pycache__ contents (#943)
* Fix #937 - Use get_data
* Add support for configuring APIGW cache TTL and encryption #942
* Addressing #909: Don't load credentials for 'package' command

## 0.42.1
* Small fixes for #918, #922, #803, #802, #799, #888, #903, #893, #828, #874, and others.
* Support for manylinux wheels Python 3.6 package downloading.
* Py3 `certify` fixes.
* Add support for multiple expressions when scheduling
* Fix content-type headers not passing through on DELETE
* Avoid creating __init__.py in a directory next to a module (.py file) with the same name
* Check recursively if there is any .py{,c} file in a directory before creating __init__.py
* Fix SNS event tasks
* Bump lambda-packages

## 0.42.0
* Cached manylinux wheel installed
* New dependency installation formatting
* Clarify "stage" vs "environment" terminology in code
* Fix problem with capitalized packages
* Delete local package if using wheels version. This saves several MBs on package size in some cases (e.g. numpy).
* Thanks to @mcrowson, @nikbora and @schuyler1d

## 0.41.3
* Various Python3 fixes
* Remove some dead code
* More binary package fixes thanks to and @nikbora and @bxm156
* Improved async tasks thanks to @schuyler1d
* Various small changes

## 0.41.2
* Support for new `lambda-packages` format (Python3 support)
* Fix `setup.py` on Windows Python3
* Fix #818 - python3 import for LE
* Support AWS-specific environment variables (with KMS)

## 0.41.1
* Add `template` command
* Add `--json` in more places
* Add `--output` to package
* Support for manylinux wheels Python 3.6 package downloading #803
from nikbora
* Fix PyEnv exit code #799

## 0.41.0
* Add Python3 Support! #793, #6
* Deprecate `lets_encrypt_expression`
* Refactor a bunch of stuff to work with Python3 package restrictions >:[
* #776 fix for libmysqlclient.so.18 error when using `slim_handler`
* add profile and region detection to init - thanks @pdpol
* #774 Wsgi environment improvements (Fix untrustworthy remote_addr)
* Only create `__init__.py` file if there are python files or sub dirs in the folder
* Update docs to reflect lambda name prepended to role_name
* Guard log responses (thanks @scoates)

## 0.40.0
* Add Async Task Support! Lots of tickets and PRs related, including #61, #603, #694 and #732.
* More info here: https://blog.zappa.io/posts/zappa-introduces-seamless-asynchronous-task-execution
* Fix Django non-WSGI function initialization, #748
* Add support for AWS Lambda Dead Letter Queue, #740
* Fix API Gateway test button (the bolt button), #735
* Switch to using per-lambda-name (project-stage) rather than a single default LambdaExecutionRole

## 0.39.1
* Fix broken Let's Encrypt trying to use new ACM ARNs
* Add `apigateway_description` setting, fixes #722
* More aggressive virtualenvironment checking

## 0.39.0
* Add `certificate_arn` setting, support for AWS Certificate Manager (#710)
* Fix zip permissions when building on Windows (#714)
* Change the active working directory to `/tmp` when using the slim handler so that relative filepaths work. (#711)

## 0.38.1
* Hotfix for broken Django deploys

## 0.38.0
* Add confirm to `certify`
* Add `--manual` to `--certify`
* Fix `certify` for existing domains
* Add `extra_permissions` setting
* Add `shell` command

## 0.37.2
* Revert to Kappa 0.6.0 #684 and others
* Add binary support for more HTTP methods, #696

## 0.37.1
* Add binary upload support, fix #683

## 0.37.0
* Add support for custom, non-Let's Encrypt certificates, thanks to Benjamin Congdon
* Change default permissions to allow executable binaries, #682
* Fix binary support for Django POST, #677

## 0.36.1
* Remove Kappa 0.6 specific hack
* Bring back '-' substitution

## 0.36.0

zappa/settings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import collections
import copy
import json
import os
import random
import string

from munch import Munch
from six import iteritems

__all__ = ['Settings']


SETTINGS_DEFAULT = {
    'api_stage': None,
    'api_key_required': False,
    'project_name': None,
    'lambda_name': None,
    's3_bucket': "zappa-" + ''.join(
                random.choice(string.ascii_lowercase + string.digits) for _ in range(9)),
    'vpc_config': {},
    'memory_size': 512,
    'app_function': None,
    'aws_region': 'us-east-1',
    'debug': True,
    'prebuild_script': None,
    'profile_name': None,
    'log_level': "DEBUG",
    'domain': None,
    'timeout_seconds': 30,
    'use_apigateway': True,
    'integration_content_type_aliases': {},
    'lambda_handler': 'handler.lambda_handler',
    'remote_env_bucket': None,
    'remote_env_file': None,
    'settings_file': None,
    'django_settings': None,
    'manage_roles': True,
    'cloudwatch_log_level': 'OFF',
    'cloudwatch_data_trace': False,
    'cloudwatch_metrics_enabled': False,
    'touch': True,
    'delete_zip': True,
    'keep_warm': True,
    'keep_warm_expression': "rate(5 minutes)",
}


class Settings(Munch):
    """

    """

    @classmethod
    def from_file(cls, filename):
        """ Return a new Settings instance from a JSON file.
            Arguments:
                filename  : File name to read.

            All open() and json.load() exceptions are propagated.
        """
        settings = cls()
        if os.path.exists(filename):
            settings.load_json(filename)
        else:
            raise RuntimeError("Settings file does not exist: {}".format(filename))
        return settings

    def load_json(self, filename=None):
        """ Load this dict from a JSON file.
            Raises the same errors as open() and json.load().
        """
        if filename or not self.getattr(self, '_settings_filename', None):
            self._settings_filename = filename

        if not self._settings_filename:
            raise ValueError('`_settings_filename` must be set.')

        with open(self._settings_filename, 'r') as f:
            data = json.load(f)

        if data is None:
            # JSON null.
            data = {}

        if not isinstance(data, dict):
            raise TypeError(
                'Data was replace with non dict type, got: {}'.format(
                    type(data)))

        self.deep_update(data)

    def save_json(self, filename=None, sort_keys=False):
        """ Save this dict to a JSON file.
            Raises the same errors as open() and json.dump().
        """
        if filename or not getattr(self, '_settings_filename', None):
            self._settings_filename = filename

        if not self._settings_filename:
            raise ValueError('`filename` must be set.')

        with open(self._settings_filename, 'w') as f:
            f.write(self.toJSON(indent=4, sort_keys=sort_keys))

    def deep_update(self, other):
        _deep_update(self, other)


def toJSON(self, **options):
    """
    Overrides the builtin Munch toJSON to remove attributes that begin with an '_' from the output
    """
    out = copy.deepcopy(self.toDict())
    for key in out.keys():
        if key.startswith('_'):
            del out[key]
    return json.dumps(out, **options)


Settings.toJSON = toJSON


def _deep_update(source,  overrides):
    """Update source with a nested dictionary or similar mapping.
    Modifies in place.
    """
    # http://stackoverflow.com/questions/3232943/update-value-of-a-nested-dictionary-of-varying-depth  # noqa
    for key, value in iteritems(overrides):
        if isinstance(value, collections.Mapping):
            if isinstance(value, collections.Mapping):
                returned = _deep_update(source.get(key, Munch()), value)
                source[key] = returned
            else:
                source[key] = overrides[key]
        else:
            source[key] = overrides[key]
    return source
